src/libstd/path.rs

```diff
@@ -1419,20 +1423,28 @@ impl<'a, T: ?Sized + AsRef<OsStr>> From<&'a T> for PathBuf {
 
 #[stable(feature = "rust1", since = "1.0.0")]
 impl From<OsString> for PathBuf {
+    /// Converts a `OsString` into a `PathBuf`.
+    /// This conversion copies the data.
+    /// This conversion does allocate memory.
     fn from(s: OsString) -> PathBuf {
         PathBuf { inner: s }
```

Comment:
Interestingly, this does not allocate memory. This is free.

`PathBuf` is a wrapper around `OsString`. If you have a `struct` with just one field, its memory structure is the same as the field itself. The Compiler "sees" through this and knows that turning OsString into a PathBuf is effectively a no-op.

See https://godbolt.org/g/1PBCTY (using `String`, as an example, but the principle is the same). If you have a close look at the blue and orange part, this just sets up for returning the previously allocated `String`. It doesn't even call the implementation!

Reply:
Thanks a lot for the explanation, was a while since I looked at Assembler. Do you have a general set of tools you use for investigating memory allocation? I sort of did all over the place with no real plan.